drivers/sensor/stmemsc/stmemsc_i2c.c

```diff
@@ -19,6 +20,12 @@ int stmemsc_i2c_read(const struct stmemsc_cfg_i2c *stmemsc,
 int stmemsc_i2c_write(const struct stmemsc_cfg_i2c *stmemsc,
 			      uint8_t reg_addr, uint8_t *value, uint8_t len)
 {
-	return i2c_burst_write(stmemsc->bus, stmemsc->i2c_slv_addr,
-			       reg_addr, value, len);
+	/* Largest stmemsc driver write is 9 bytes */
+	uint8_t buffer[10];
+
+	__ASSERT((1U + len) <= sizeof(buffer), "stmemsc i2c buffer too small");
+	buffer[0] = reg_addr;
+	memcpy(buffer + 1, value, len);
```

Comment:
__ASSERTs are useless. `memcpy(&buffer[1], value, MIN(len, sizeof(buffer) - 1));` ?

Reply:
But here you should at least report that you are writing less than requested. Something wrong will happen somewhere in any case. Moreover, as I said earlier, I feel that imposing a maximum size (reg + 9 byte of data here) may lead to issues in the next future.

Reply:
Don't use a literal 9 or 10; define a constant in a header that specifies the maximum data length for one of the relevant operations (e.g. maximum size of config structures, if those are the ones that are a problem), and document it needs to be updated when things change.

If you put that define in the upstream stmemsc package then everything should be good because the stmemsc maintainers should know when new sensors that have larger transfers are added.

Reply:
@pabigot
I am the stmemsc maintainer (sigh!).
The issue here is that things may get more complex in the future, with sensor that has the capability to drive external sensors (the shub feature is already there in some of the current sensor drivers) and maybe other things. Calculating the maximum size and adapting the macro evrytime a new sensor (or a new feature in a sensor) can lead to trivial errors. 
Frankly speaking I would rather keep the i2c_burst_write() for the SoC that have I2C scatter-gather capability and the i2c_write implementation for the others. If we want to avoid the NFRX platform test, maybe we can add in stmemsc a new property that provides the burst-skipping info (and maybe another with the max buffer size).
In this way, we confine the work to stmemsc w/o affecting the I2C TWIM driver behaviour. Of course the default of that property will be "use-burst". The NFRx platforms will have to indicate the alternative usage.
What do you all think?

Reply:
Something like "stmemsc,buf-size". When is 0 (the default) means go with i2c_burst_write. The property must be included in stmemsc based sensors when they are on a NFRx platform.

Reply:
Responding only because I was directly asked for this: I don't have a horse in this race because I've given up on Zephyr being useful to me personally, so my patience with [wicked issues](https://en.wikipedia.org/wiki/Wicked_problem) like this is exhausted.

There are so many ways this could be resolved; each has different trade-offs of (a) amount of cooperation required from reviewers and other maintainers to get the solution merged, and (b) impact on Zephyr maintainability and usability.

I don't see where `stmemsc,buf-size` could be provided or how its value could be determined by most users, so I would do this:

```
#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_twim)
  /* The underlying I2C driver might be nordic,nrf-twim, and if so it might
   * not have a zephyr,concat-buf-size large enough to correctly implement
   * i2c_burst_write of len bytes.  Since we can't verify that easily at
   * either build or run time, pack the write payload locally.
   */
  uint8_t buf[1 + len];
  
  buf[0] = reg_addr;
  memcpy(buf+1, value, len);
  return i2c_write(...);
#else
  return i2c_burst_write(...);
#endif
```

This is conservative in that on Nordic platforms it sometimes uses non-burst write in cases where burst-write would have worked, but checking for that at build time (or even at runtime) is IMO not feasible or doesn't solve the problem of clearly indicating to the user that there's a knob that could be set to make the code work.

But depending on Zephyr coding guidelines the use of a VLA might be rejected, in which case the alternative is a constant that comes from a header file or Kconfig and a runtime check against length.  But that still wouldn't solve the problem of a novice user wondering why their sensor doesn't work.

Reply:
> Responding only because I was directly asked for this: I don't have a horse in this race because I've given up on Zephyr being useful to me personally, so my patience with [wicked issues](https://en.wikipedia.org/wiki/Wicked_problem) like this is exhausted.
> 
Sorry for that.

> There are so many ways this could be resolved; each has different trade-offs of (a) amount of cooperation required from reviewers and other maintainers to get the solution merged, and (b) impact on Zephyr maintainability and usability.
> 
> I don't see where `stmemsc,buf-size` could be provided or how its value could be determined by most users, so I would do this:
> 
> ```
> #if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_twim)
I think you need a DT_DRV_COMPAT to do that, but we do not have in stmemsc as this is more like a library used by many drivers. I would prefer the original check as before (i.e. IS_ENABLED(CONFIG_I2C_NRFX)

>   /* The underlying I2C driver might be nordic,nrf-twim, and if so it might
>    * not have a zephyr,concat-buf-size large enough to correctly implement
>    * i2c_burst_write of len bytes.  Since we can't verify that easily at
>    * either build or run time, pack the write payload locally.
>    */
>   uint8_t buf[1 + len];
I think we should avoid arrays with variable length, and accept the fact that, in case of nfrx TWIM platforms, we will handle the normal 9axis case (9 + 1 byte) as it was before.

More complex sensor features will not be handled here.

>   
>   buf[0] = reg_addr;
>   memcpy(buf+1, value, len);
>   return i2c_write(...);
> #else
>   return i2c_burst_write(...);
> #endif
> ```
> 
> This is conservative in that on Nordic platforms it sometimes uses non-burst write in cases where burst-write would have worked, but checking for that at build time (or even at runtime) is IMO not feasible or doesn't solve the problem of clearly indicating to the user that there's a knob that could be set to make the code work.
> 
> But depending on Zephyr coding guidelines the use of a VLA might be rejected, in which case the alternative is a constant that comes from a header file or Kconfig and a runtime check against length. But that still wouldn't solve the problem of a novice user wondering why their sensor doesn't work.

In my opinion it would be better to avoid VLAs.